File: TS/系统/03．技能系统/00．技能模板+函数/00．技能模板/08．装备触发模板/00．击杀回复触发模板.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;
const japi = require("jass.japi") as any;

const { registerDeathListener } = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心") as {
  registerDeathListener: (this: void, callback: (this: void, dyingUnit: any, killingUnit: any) => void) => void;
};
const { getServerTime } = require("系统.00．核心系统.05．中心计时器") as {
  getServerTime: (this: void) => number;
};
const { doHeal } = require("系统.04．伤害系统.02．治疗系统.01．核心功能") as {
  doHeal: (this: void, params: {
    HealSource: any;
    HealTarget: any;
    HealAmount: number;
    HealManaAmount?: number;
    ItemHeal: boolean;
    HealEffect: boolean;
    HealEffectPath?: string;
    UseDefaultHealEffect?: boolean;
    ManaEffect?: boolean;
    ManaEffectPath?: string;
    UseDefaultManaEffect?: boolean;
    ManaShowText?: boolean;
    DelayOneTick?: boolean;
  }) => number;
};

const GetHandleId = jass.GetHandleId as (handle: any) => number;
const GetUnitState = jass.GetUnitState as (unit: any, state: any) => number;
const IsUnitType = jass.IsUnitType as (unit: any, whichType: any) => boolean;
const IsUnitEnemy = jass.IsUnitEnemy as (whichUnit: any, whichPlayer: any) => boolean;
const GetOwningPlayer = jass.GetOwningPlayer as (whichUnit: any) => any;
const UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD as any;
const UNIT_STATE_MAX_LIFE = jass.UNIT_STATE_MAX_LIFE as any;
const UNIT_STATE_MAX_MANA = jass.UNIT_STATE_MAX_MANA as any;
const GetUnitStateJapi = japi.GetUnitState as (unit: any, state: any) => number;

export interface 击杀回复触发事件 {
  击杀单位: any;
  死亡单位: any;
  配置: 击杀回复触发模板配置;
}

export interface 击杀回复触发模板配置 {
  名称?: string;
  冷却秒数?: number;
  恢复生命值?: number;
  恢复最大生命比例?: number;
  恢复魔法值?: number;
  恢复最大魔法比例?: number;
  只触发敌方死亡?: boolean;
  触发条件?: (this: void, event: 击杀回复触发事件) => boolean;
  生命特效路径?: string;
  魔法特效路径?: string;
  使用默认生命特效?: boolean;
  使用默认魔法特效?: boolean;
  on触发后?: (this: void, event: 击杀回复触发事件, 恢复生命: number, 恢复魔法: number) => void;
}

export interface 击杀回复触发控制器 {
  readonly 名称: string;
  停止(this: void): void;
}

interface 击杀回复触发记录 extends 击杀回复触发控制器 {
  ID: number;
  配置: 击杀回复触发模板配置;
  已停止: boolean;
}

const 击杀回复触发记录表: Record<number, 击杀回复触发记录 | undefined> = {};
const 击杀回复冷却表: Record<string, number | undefined> = {};
let 击杀回复触发计数 = 0;
let 已注册击杀回复死亡监听 = false;

function 单位有效存活(this: void, unit: any): boolean {
  return unit != null && unit !== 0 && IsUnitType(unit, UNIT_TYPE_DEAD) !== true && GetUnitState(unit, jass.UNIT_STATE_LIFE) > 0.405;
}

function 取单位ID(this: void, unit: any): number {
  if (unit == null || unit === 0) return 0;
  return GetHandleId(unit) || 0;
}

function 取最大生命(this: void, unit: any): number {
  return GetUnitStateJapi(unit, UNIT_STATE_MAX_LIFE) || GetUnitState(unit, UNIT_STATE_MAX_LIFE) || 0;
}

function 取最大魔法(this: void, unit: any): number {
  return GetUnitStateJapi(unit, UNIT_STATE_MAX_MANA) || GetUnitState(unit, UNIT_STATE_MAX_MANA) || 0;
}

function 取冷却键(this: void, unit: any, record: 击杀回复触发记录): string {
  const id = 取单位ID(unit);
  if (id === 0) return "";
  return String(record.ID) + ":" + String(id);
}

function 冷却通过并记录(this: void, unit: any, record: 击杀回复触发记录): boolean {
  const cd = record.配置.冷却秒数 ?? 0;
  if (!(cd > 0)) return true;
  const key = 取冷却键(unit, record);
  if (key === "") return false;
  const now = getServerTime();
  const end = 击杀回复冷却表[key];
  if (end != null && now < end) return false;
  击杀回复冷却表[key] = now + cd * 1000;
  return true;
}

function 计算恢复生命(this: void, unit: any, config: 击杀回复触发模板配置): number {
  return (config.恢复生命值 ?? 0) + 取最大生命(unit) * (config.恢复最大生命比例 ?? 0);
}

function 计算恢复魔法(this: void, unit: any, config: 击杀回复触发模板配置): number {
  return (config.恢复魔法值 ?? 0) + 取最大魔法(unit) * (config.恢复最大魔法比例 ?? 0);
}

function 尝试执行击杀回复(this: void, dyingUnit: any, killingUnit: any, record: 击杀回复触发记录): void {
  if (record.已停止 || !单位有效存活(killingUnit)) return;
  if (dyingUnit == null || dyingUnit === 0 || dyingUnit === killingUnit) return;
  if (record.配置.只触发敌方死亡 !== false && !IsUnitEnemy(dyingUnit, GetOwningPlayer(killingUnit))) return;

  const event: 击杀回复触发事件 = { 击杀单位: killingUnit, 死亡单位: dyingUnit, 配置: record.配置 };
  if (record.配置.触发条件 != null && !record.配置.触发条件(event)) return;
  if (!冷却通过并记录(killingUnit, record)) return;

  const heal = 计算恢复生命(killingUnit, record.配置);
  const mana = 计算恢复魔法(killingUnit, record.配置);
  if (!(heal > 0) && !(mana > 0)) return;

  doHeal({
    HealSource: killingUnit,
    HealTarget: killingUnit,
    HealAmount: heal,
    HealManaAmount: mana,
    ItemHeal: true,
    HealEffect: record.配置.使用默认生命特效 === true || (record.配置.生命特效路径 != null && record.配置.生命特效路径 !== ""),
    HealEffectPath: record.配置.生命特效路径,
    UseDefaultHealEffect: record.配置.使用默认生命特效 === true,
    ManaEffect: record.配置.使用默认魔法特效 === true || (record.配置.魔法特效路径 != null && record.配置.魔法特效路径 !== ""),
    ManaEffectPath: record.配置.魔法特效路径,
    UseDefaultManaEffect: record.配置.使用默认魔法特效 === true,
    ManaShowText: mana > 0,
  });

  if (record.配置.on触发后 != null) record.配置.on触发后(event, heal, mana);
}

function on击杀回复死亡事件(this: void, dyingUnit: any, killingUnit: any): void {
  for (const key in 击杀回复触发记录表) {
    const record = 击杀回复触发记录表[Number(key) || 0];
    if (record != null) 尝试执行击杀回复(dyingUnit, killingUnit, record);
  }
}

function 确保击杀回复死亡监听(this: void): void {
  if (已注册击杀回复死亡监听) return;
  已注册击杀回复死亡监听 = true;
  registerDeathListener(on击杀回复死亡事件);
}

export function 注册击杀回复触发模板(this: void, 配置: 击杀回复触发模板配置): 击杀回复触发控制器 {
  确保击杀回复死亡监听();
  const id = ++击杀回复触发计数;
  const record: 击杀回复触发记录 = {
    ID: id,
    名称: 配置.名称 ?? ("击杀回复触发#" + String(id)),
    配置,
    已停止: false,
    停止: function 停止击杀回复触发(this: void): void {
      record.已停止 = true;
      delete 击杀回复触发记录表[id];
    },
  };
  击杀回复触发记录表[id] = record;
  return record;
}

export {};

File: src/系统/03．技能系统/00．技能模板+函数/00．技能模板/08．装备触发模板/00．击杀回复触发模板.lua
local ____lualib = require("lualib_bundle")
local __TS__Number = ____lualib.__TS__Number
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local japi = require("jass.japi")
local ____require_result_0 = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心")
local registerDeathListener = ____require_result_0.registerDeathListener
local ____require_result_1 = require("系统.00．核心系统.05．中心计时器")
local getServerTime = ____require_result_1.getServerTime
local ____require_result_2 = require("系统.04．伤害系统.02．治疗系统.01．核心功能")
local doHeal = ____require_result_2.doHeal
local GetHandleId = jass.GetHandleId
local GetUnitState = jass.GetUnitState
local IsUnitType = jass.IsUnitType
local IsUnitEnemy = jass.IsUnitEnemy
local GetOwningPlayer = jass.GetOwningPlayer
local UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD
local UNIT_STATE_MAX_LIFE = jass.UNIT_STATE_MAX_LIFE
local UNIT_STATE_MAX_MANA = jass.UNIT_STATE_MAX_MANA
local GetUnitStateJapi = japi.GetUnitState
local _____51FB_6740_56DE_590D_89E6_53D1_8BB0_5F55_8868 = {}
local _____51FB_6740_56DE_590D_51B7_5374_8868 = {}
local _____51FB_6740_56DE_590D_89E6_53D1_8BA1_6570 = 0
local _____5DF2_6CE8_518C_51FB_6740_56DE_590D_6B7B_4EA1_76D1_542C = false
local function _____5355_4F4D_6709_6548_5B58_6D3B(unit)
    return unit ~= nil and unit ~= 0 and IsUnitType(unit, UNIT_TYPE_DEAD) ~= true and GetUnitState(unit, jass.UNIT_STATE_LIFE) > 0.405
end
local function _____53D6_5355_4F4DID(unit)
    if unit == nil or unit == 0 then
        return 0
    end
    return GetHandleId(unit) or 0
end
local function _____53D6_6700_5927_751F_547D(unit)
    return GetUnitStateJapi(unit, UNIT_STATE_MAX_LIFE) or GetUnitStateJapi(unit, UNIT_STATE_MAX_LIFE) or 0
end
local function _____53D6_6700_5927_9B54_6CD5(unit)
    return GetUnitStateJapi(unit, UNIT_STATE_MAX_MANA) or GetUnitStateJapi(unit, UNIT_STATE_MAX_MANA) or 0
end
local function _____53D6_51B7_5374_952E(unit, record)
    local id = _____53D6_5355_4F4DID(unit)
    if id == 0 then
        return ""
    end
    return (tostring(record.ID) .. ":") .. tostring(id)
end
local function _____51B7_5374_901A_8FC7_5E76_8BB0_5F55(unit, record)
    local cd = record["配置"]["冷却秒数"] or 0
    if not (cd > 0) then
        return true
    end
    local key = _____53D6_51B7_5374_952E(unit, record)
    if key == "" then
        return false
    end
    local now = getServerTime()
    local ____end = _____51FB_6740_56DE_590D_51B7_5374_8868[key]
    if ____end ~= nil and now < ____end then
        return false
    end
    _____51FB_6740_56DE_590D_51B7_5374_8868[key] = now + cd * 1000
    return true
end
local function _____8BA1_7B97_6062_590D_751F_547D(unit, config)
    return (config["恢复生命值"] or 0) + _____53D6_6700_5927_751F_547D(unit) * (config["恢复最大生命比例"] or 0)
end
local function _____8BA1_7B97_6062_590D_9B54_6CD5(unit, config)
    return (config["恢复魔法值"] or 0) + _____53D6_6700_5927_9B54_6CD5(unit) * (config["恢复最大魔法比例"] or 0)
end
local function _____5C1D_8BD5_6267_884C_51FB_6740_56DE_590D(dyingUnit, killingUnit, record)
    if record["已停止"] or not _____5355_4F4D_6709_6548_5B58_6D3B(killingUnit) then
        return
    end
    if dyingUnit == nil or dyingUnit == 0 or dyingUnit == killingUnit then
        return
    end
    if record["配置"]["只触发敌方死亡"] ~= false and not IsUnitEnemy(
        dyingUnit,
        GetOwningPlayer(killingUnit)
    ) then
        return
    end
    local event = {["击杀单位"] = killingUnit, ["死亡单位"] = dyingUnit, ["配置"] = record["配置"]}
    if record["配置"]["触发条件"] ~= nil and not record["配置"]["触发条件"](event) then
        return
    end
    if not _____51B7_5374_901A_8FC7_5E76_8BB0_5F55(killingUnit, record) then
        return
    end
    local heal = _____8BA1_7B97_6062_590D_751F_547D(killingUnit, record["配置"])
    local mana = _____8BA1_7B97_6062_590D_9B54_6CD5(killingUnit, record["配置"])
    if not (heal > 0) and not (mana > 0) then
        return
    end
    doHeal({
        HealSource = killingUnit,
        HealTarget = killingUnit,
        HealAmount = heal,
        HealManaAmount = mana,
        ItemHeal = true,
        HealEffect = record["配置"]["使用默认生命特效"] == true or record["配置"]["生命特效路径"] ~= nil and record["配置"]["生命特效路径"] ~= "",
        HealEffectPath = record["配置"]["生命特效路径"],
        UseDefaultHealEffect = record["配置"]["使用默认生命特效"] == true,
        ManaEffect = record["配置"]["使用默认魔法特效"] == true or record["配置"]["魔法特效路径"] ~= nil and record["配置"]["魔法特效路径"] ~= "",
        ManaEffectPath = record["配置"]["魔法特效路径"],
        UseDefaultManaEffect = record["配置"]["使用默认魔法特效"] == true,
        ManaShowText = mana > 0
    })
    if record["配置"]["on触发后"] ~= nil then
        record["配置"]["on触发后"](event, heal, mana)
    end
end
local function ____on_51FB_6740_56DE_590D_6B7B_4EA1_4E8B_4EF6(dyingUnit, killingUnit)
    for key in pairs(_____51FB_6740_56DE_590D_89E6_53D1_8BB0_5F55_8868) do
        local record = _____51FB_6740_56DE_590D_89E6_53D1_8BB0_5F55_8868[__TS__Number(key) or 0]
        if record ~= nil then
            _____5C1D_8BD5_6267_884C_51FB_6740_56DE_590D(dyingUnit, killingUnit, record)
        end
    end
end
local function _____786E_4FDD_51FB_6740_56DE_590D_6B7B_4EA1_76D1_542C()
    if _____5DF2_6CE8_518C_51FB_6740_56DE_590D_6B7B_4EA1_76D1_542C then
        return
    end
    _____5DF2_6CE8_518C_51FB_6740_56DE_590D_6B7B_4EA1_76D1_542C = true
    registerDeathListener(____on_51FB_6740_56DE_590D_6B7B_4EA1_4E8B_4EF6)
end
____exports["注册击杀回复触发模板"] = function(_____914D_7F6E)
    _____786E_4FDD_51FB_6740_56DE_590D_6B7B_4EA1_76D1_542C()
    _____51FB_6740_56DE_590D_89E6_53D1_8BA1_6570 = _____51FB_6740_56DE_590D_89E6_53D1_8BA1_6570 + 1
    local id = _____51FB_6740_56DE_590D_89E6_53D1_8BA1_6570
    local record
    record = {
        ID = id,
        ["名称"] = _____914D_7F6E["名称"] or "击杀回复触发#" .. tostring(id),
        ["配置"] = _____914D_7F6E,
        ["已停止"] = false,
        ["停止"] = function()
            record["已停止"] = true
            __TS__Delete(_____51FB_6740_56DE_590D_89E6_53D1_8BB0_5F55_8868, id)
        end
    }
    _____51FB_6740_56DE_590D_89E6_53D1_8BB0_5F55_8868[id] = record
    return record
end
return ____exports
